### ISSUE TITLE
Strange behavior of param type match

### ISSUE THREAD

**Issue by MillhioreBT** [EmmyLuaLs/emmylua-analyzer-rust#458]
![Image](https://github.com/user-attachments/assets/0baa0b3d-9f2f-4bc5-aa39-d82d2a9beac5)

![Image](https://github.com/user-attachments/assets/c93ae97a-433c-48c5-9a8c-3e7843135a87)

```lua
---@generic K, V
---@param luaTable table<K, V> | V[]
---@param value V
---@return boolean
function table.contains(luaTable, value)
	for _, targetColumn in pairs(luaTable) do if targetColumn == value then return true end end
	return false
end
```

This simply should not throw any error or warning, since it should just accept any numeric value
it's an inline table and not an enum.

**Comment by CppCXY** [comment]
because `{ 1758, 5502 }`  is prioritized as a tuple. I suggest adding --[[@as integer[] ]]

**Comment by taminomara** [comment]
I believe we could be less strict in these cases, although I'm not sure how to implement this.

If we merge #457, there will be another similar inconvenience:

```lua
local arr = { 1, 2, 3 }
table.insert(arr, 4) --> Warning: can't insert `4` into `(1|2|3)[]`.
```

**Comment by MillhioreBT** [comment]
> I believe we could be less strict in these cases, although I'm not sure how to implement this.
> 
> If we merge [#457](https://github.com/EmmyLuaLs/emmylua-analyzer-rust/pull/457), there will be another similar inconvenience:
> 
> local arr = { 1, 2, 3 }
> table.insert(arr, 4) --> Warning: can't insert `4` into `(1|2|3)[]`.

it's true, Lua is dynamic and making typing too strict leads to bad programming practices, such as adding unnecessary checks, disabling too many diagnostics per line, etc...

by default, it would be better to consider the generic table type or the types within the generic table rather than only accepting the set of values defined in the table.

**Comment by xuhuanzy** [comment]
In this case, the table defaults to a tuple, which is incorrect. However, unfortunately, after discussion, the leader cppcxy disagreed with making the change, even though it is wrong. Therefore, this issue will not be completely resolved.

**Comment by xuhuanzy** [comment]
The reason cppcxy disagreed is: "This is how it is now, and for the sake of compatibility, we have to accept it."  
However, whether this feature has such a broad impact is questionable, but in the end, the discussion concluded with him refusing to make the change.

**Comment by MillhioreBT** [comment]
> The reason cppcxy disagreed is: "This is how it is now, and for the sake of compatibility, we have to accept it." However, whether this feature has such a broad impact is questionable, but in the end, the discussion concluded with him refusing to make the change.

Yes, I perfectly understand that changing this would break many tests
I think the correct option is to define in the configuration whether we want the tables to be interpreted as tuples in the generics

**Comment by taminomara** [comment]
We can benefit from a smarter algorithm that infers generic parameters from arguments. Right now, we match argument types against ones specified in documentation, and the first type that matches wins. So, for this example:

```lua
table.contains({ 1, 2, 3 }, 4)
```

we will match `{ 1, 2, 3 }` against `T`, infer that `T[]` is `1|2|3`, then we discard the info that the second parameter is `4`.

Instead, we can match all two parameters, and gather possible types. So, we will know that `T` is `1|2|3` or `4`. From there, we can try to infer the lowest common type for `T`, which in this case will be `integer` (or maybe `1|2|3|4`, depending on implementation).

This way, we will achieve flexibility without breaking current tests.

**Comment by taminomara** [comment]
For the case of

```lua
local arr = { 1, 2, 3 }
table.insert(arr, 4)
```

this will not work. However, I believe a more common case is

```lua
local arr = {}
table.insert(arr, 4)
```

which works now.

**Comment by taminomara** [comment]
We will have to be careful around variance, though. For example, we can't cast a tuple of two integers into an array of integers, because we don't want `table.insert(tuple, x)` to work.

**Comment by MillhioreBT** [comment]
I'm going to close it because this change is not really necessary since it can be handled with the `any` type, anyway this method is quite generic on its own, `any` is the correct solution for my use case
